### ISSUE TITLE
[Feature/Bug?]: Cache models to RAM.

### ISSUE THREAD

**Issue by midcoastal** [vladmandic/automatic#2178]
### Feature description

Loving the new optimizations, and would like to request another. ;-)

It would be extremely helpful if there were an option to cache N recently-used models to RAM (not VRAM), to speed up loading.  Particularly when using a refiner.

I do note that the are options `Move X to CPU when not in use.` etc, but that doesn't seem to behave as expected?  When I have a refiner selected, watching my GPU memory usage, I can see that both models get loaded (and stay loaded) into VRAM at load, which, needless to say, doesn't end well.  

I should probably additionally note that this is with the same model selected for the base and refiner.  Which, incidentally, raises another question: 

- "If they are the same, why not just use the one model and not load an additional?"

Beyond just the base and refiner, however, it would be helpful to have the option to have a cache of N models in RAM, for faster load times.  I have 128gb, and it would be super awesome if `adetailer` didn't have to load the model from disk on every image (I am using separate models for detailing).

And, going even a step further (I know, I am sorry), with this cacheing, it would make sense to only apply xformers/sdp optimizations once, cache the result, and load that from RAM to save even more time.  But I know that is perhaps WAY too much.  I'll only ask for the RAM cacheing for now.  ;-)

**Comment by midcoastal** [comment]
Waiiiit a tick...  I was able to get my RAM usage up to like 400gb (including swap, not ideal, but swap is on an NVMe raid, so better than the bulk storage the models are on), but I can still see I/O loading models from storage...

When using the model cache, how does it determine which models to cache?  I am guessing it isn't going by use?, otherwise, I only have 6 models I am using in this rotation, and even if we just say they are 20gb a pop (they most certainly are not) that is only 120gb.  I am using 3 Loras, let's just say 10gb (again, just using arbitrarily large numbers), still only 150gb.

So...  what, exactly, is it cacheing?  lol

**Comment by vladmandic** [comment]
its evaluated at model load and then looks as FIFO algorithm - first-in-first-out while keeping n last models in ram.
but that code is quite old, so seems like you're seeing a massive leak.
i'd need to take a much deeper look at this, placing in backlog.

**Comment by midcoastal** [comment]
Ok, this is turning into a bug report:

- I can confirm it is reusing models in RAM cache (multiple runs on a fresh load, no I/O to storage).
- I believe it is also moving the models *back* into RAM (instead of *not* if the model already exists in RAM).  I believe this because:
  - I can see my ram usage tick up after each render, by *about* the size of a model (I am unsure **which** model, as as I said, I am using `adetailer` with separate models, a refiner, and some VAEs, *however* best I can tell, it is not "all" models, but just one or two of the several I am using).
  - Looking at timing, it seems it **may** be the `adetailer` models, as it seems to tick up when it is loading it's models (but unsure if that is a move of a standard model back in to ram, or loading the `adetailer` additionally into ram *again*).
  - I can see several places where it states `from cache`, it would be **very** helpful to also have indicators for `writing to cache`, `removing from cache`, `leaving in cache`, and `skipping cache write (exists)`.
  - Without looking at the inner workings, I do know something isn't behaving correctly, as when the cache count is full, it seems to drop an actively used model out of cache, as the eviction results in a re-load of a model from disk.  When running in a batch, one would assume that there are "hot" models that should not be evicted, and even if "extraneous" models/copies were being written to RAM, it should be the eldest, *not used* extra copies that were getting evicted, but that does not seem to be the case.  It doesn't even look like a circular cache, as a "relevant" model seems to get evicted and re-loaded on each batch.

Over all, very strange behavior.  On the other hand, it is nice to see all of my RAM being used, for once.  It took me the LONGEST time to get a stable DDR5 tune on 4 sticks, and day to day I only use 25% or so.

Oh, also, it does seem that it may be popping models off the stack on load, as the ticks down ever so slightly as well, but obviously up, more than down.

If it is FIFO, it seems there is a leak somewhere, and crap is staying behind, making the cache useless after a period of time...

**Comment by midcoastal** [comment]
> its evaluated at model load and then looks as FIFO algorithm - first-in-first-out while keeping n last models in ram. but that code is quite old, so seems like you're seeing a massive leak. i'd need to take a much deeper look at this, placing in backlog.

**BACKLOG???**  Sad days, my man...

Hey, I'll trade you: When this is patched, I will try my hand again at that cacheing bit, OR (since the whole concept of how the Diffusers directories really sets my teeth on edge, and your patches are super good as is) a JS implementation of Extra networks UI.  ;-)

**Comment by vladmandic** [comment]
yeah, it does sound like a bug. i'm placing in backlog only because that code is legacy as hell, i should pretty much rewrite it instead of trying to fix it.

regarding caching, it should be much better in the current versions - did you try?

**Comment by midcoastal** [comment]
> yeah, it does sound like a bug. i'm placing in backlog only because that code is legacy as hell, i should pretty much rewrite it instead of trying to fix it.
> 
> regarding caching, it should be much better in the current versions - did you try?

Yes, start up is so much better that I don't have time to go grab lunch when I restart the server.  Very annoying.  Only enough time to go whip up a PB&J.

That said, I was, for a good bit, getting Segfaults every which way to Sunday.  I was using kernel 6.3, tried 6.4 and 6.5, but the NVidia drivers are bugged for those, tried 525 and 535 on 6.3 and 6.2.  Currently running 535 on Canonical's `6.2.0-32.32~22.04.1`.  Less than ideal, as I would at least like to be running the latest 6.2.16, and ideally minimum of 6.3.13 (best case latest 6.4), but whatever, it isn't crashing all the time.

I tried a new venv, and catching segfaults with the fault handler and 3.10-dbg, but the stack wasn't consistent, and nothing really stood out.  I may try again.  The one consistent was a thread that was always somewhere inside the Torch module.  We'll see.

As for the mem leak...  You're really going to make me do it, aren't you.  ;-)  Point a finger at a starting point, I'll poke it.

**Comment by midcoastal** [comment]
**Lol...  Side note:**

I have to apply a standing local patch in the `gradio` module in order for WebUI to successfully launch.  Evidently, the main thread is doing something with models/thumbnails (or is at least affected by thread locks) around the same time the Gradio module does it's startup health-check.  Causes the server to be sluggish, and the default request timeout of 3 seconds isn't long enough.  I have to set it to 5 or better to successfully load.

This was the first patch outside of the app repo I have had to do.  Looks like I found an upper limit on model counts...  ;-)

**Comment by vladmandic** [comment]
settings -> extra networks -> do not automatically build extra network pages
(result is that en will be empty until you manually click on refresh)

**Comment by midcoastal** [comment]
> * settings -> extra networks -> do not automatically build extra network pages
>   result is that en will be empty until you manually click on refresh
> * `webui --no-metadata`
>   to skip reading of model metadata on server startup

Will the metadata be read when referenced?  If that how it works?

**Comment by vladmandic** [comment]
In the first case, yes.
In the second, no. It's just disabled.

**Comment by midcoastal** [comment]
The memory leak seems to not be an issue at this time.  I have been using the caching mech since this was originally posted, and things seem to behave nominally at this time.  This could probably be put on a timeout at this time (kill it after a time/shortly).
